### ISSUE TITLE
Asign variables through other variables

### ISSUE THREAD

**Issue by Paultje52** [nodejs/help#2260]
* **Version**: `v10.16.0`
* **Platform**: `Windows 64-bit`
* **Subsystem**: ???

I can change the value of a key in an object while using another variable.

So, I have a simple object here:
```js
let object = {
  foo: "bar"
};
```
And, when I want to make another object with the same key(s), I do this.
```js
let secondObject = object;
```
So, when I console.log `secondObject`, I get `{ foo: 'bar' }`. But now I want to change the second object without changing the first object:
```js
secondObject.foo = "bar2";
```
But, when I now log the first object, I get this: `{ foo: 'bar2' }`, but I didn't change the first object. This doesn't make any sense. 

It even works through more than one object:
```js
let object = {foo: "bar"}
let secondObject = object;
let thirdObject = secondObject;
console.log(object); // { foo: 'bar' }
thirdObject.foo = 1;
console.log(object); // { foo: 1 }
```

**Comment by addaleax** [comment]
Hi @Paultje52!

On some level, this is just how JavaScript works (and how a lot of other programming languages work) – assignments don’t copy objects, they create new references to the same object.

A quick way to copy objects is to use the spread operator:

```js
let secondObject = { ...object };
```

will create a new object (as `{}` always does) and copy all of the first object’s properties into it.

I hope this helps!

**Comment by Paultje52** [comment]
Thanks for your help!

**Comment by Paultje52** [comment]
Maybe you can also add a new function to the object prototype, named `clone`, so your code is more readable.
```js
Let secondObject = object.clone();
```

**Comment by Hakerh400** [comment]
> Maybe you can also add a new function to the object prototype

Node.js is not in a position to alter native prototypes (theoretically it can, but would be very inefficient and could be considered as a bad practice). The JavaScript interpreter that Node.js uses is V8 and it strictly follows the official EcmaScript specification. TC39 committee is the place for reporting proposals regarding specification modifications, but this particular one (adding `clone` method to the Object's prototype) is highly likely to be rejected due to several reasons.

From the theoretical point of view, programming languages can be separated into two categories based on their definition of equivalence operator: intensional and extensional programming languages.

Two entity references (e.g. objects) in an intensional programming language are considered to be equal if and only if they reference the same entity (for example the same location in the virtual memory). Assigning a value to a variable actually means assigning a reference (to the value) to the variable. Most of the imperative and object-oriented, and some functional, programming languages are intensional. Examples are: JavaScript, Java, Python, C, C++, PHP, Kotlin, GoLang, Ruby...

On the other hand, two entities in an extensional programming language are considered to be equal if and only if they exhibit the same behavior in identical contexts. References, assignments and allocations in these languages do not make sense. Examples are: Haskell, Prolog, SQL, Coq..

Back to your question. Making a copy of an object in JavaScript is not well-defined. To be precise, it is, in general, impossible. Using the syntax `a = {...b}` does not copy all of the `b`'s properties. It only copies own enumerable property values. It does not copy the prototype chain, non-enumerable properies, getters, setters, etc. Objects can also have cyclic references, native methods, instances of a cross-realm class, primitive values... Moreover, there could be Proxies, which can't be distinguished from the non-proxified objects, as seen from the JavaScript language level. There could be closures, which are impossible to inspect or copy. There could be exotic objects, like arrays. There could be Buffer instances.

In summary: it is impossible to implement a perfect function for copying JavaScript values.
